sklearn/decomposition/_dict_learning.py

```diff
@@ -1567,6 +1722,182 @@ def __init__(self, n_components=None, *, alpha=1, n_iter=1000,
+            dictionary = S[:, np.newaxis] * dictionary
+        r = len(dictionary)
+        if self._n_components <= r:
+            dictionary = dictionary[:self._n_components, :]
+        else:
+            dictionary = np.r_[dictionary,
+                               np.zeros((self._n_components - r,
+                                         dictionary.shape[1]))]
+
+        dictionary = check_array(dictionary, order='F', dtype=np.float64,
+                                 copy=False)
+        dictionary = np.require(dictionary, requirements='W')
+
+        return dictionary
+
+    def _minibatch_step(self, X, dictionary, random_state, step):
+        """The guts of the algorithm"""
+        batch_size = X.shape[0]
+
+        # Compute code for this batch
+        code = sparse_encode(
+            X, dictionary, algorithm=self._fit_algorithm,
+            alpha=self.alpha, n_jobs=self.n_jobs, check_input=False,
+            positive=self.positive_code, max_iter=self.transform_max_iter,
+            verbose=self.verbose)
+
+        batch_cost = (0.5 * ((X - code @ dictionary)**2).sum()
+                      + self.alpha * np.sum(np.abs(code))) / batch_size
+
+        # Update inner stats
+        self._update_inner_stats(X, code, batch_size, step)
+
+        # Update dictionary
+        A, B = self._inner_stats
+        _update_dict(dictionary, X, code, A, B, verbose=self.verbose,
+                     random_state=random_state, positive=self.positive_dict)
+
+        return batch_cost
+
+    def _update_inner_stats(self, X, code, batch_size, step):
+        """Update the inner stats inplace"""
+        if step < batch_size - 1:
+            theta = (step + 1) * batch_size
+        else:
+            theta = batch_size ** 2 + step + 1 - batch_size
+        beta = (theta + 1 - batch_size) / (theta + 1)
+
+        A, B = self._inner_stats
+        A *= beta
+        A += np.dot(code.T, code)
+        B *= beta
+        B += np.dot(X.T, code)
+
+    def _minibatch_convergence(self, X, batch_cost, dictionary, dict_buffer,
+                               n_samples, step, n_steps):
+        """Helper function to encapsulate the early stopping logic"""
+        batch_size = X.shape[0]
+
+        # counts steps starting from 1 for user friendly verbose mode.
+        step = step + 1
+
+        # Ignore first iteration because dictionary is not projected on the
+        # constraint set yet.
+        if step == 1:
+            if self.verbose:
+                print(f"Minibatch step {step}/{n_steps}: mean batch "
+                      f"cost: {batch_cost}")
+            return False
+
+        # Compute an Exponentially Weighted Average of the cost function to
+        # monitor the convergence while discarding minibatch-local stochastic
+        # variability: https://en.wikipedia.org/wiki/Moving_average
+        if self._ewa_cost is None:
+            self._ewa_cost = batch_cost
```

Comment:
Maybe we could have a burnin phase where all the batch costs are recorded in a numpy arrays of 100 batches for instance and then initialized `self._ewa_cost = np.mean(self._ewa_cost_burnin)` to avoid being too much impacted by the randomness of the cost of the first batches?

Reply:
done. I did 100 steps or 1 epoch